### ISSUE TITLE
[Bug]: FlashInfer attention + FP8 KV cache + CUDA graphs produces random output on RTX 6000 Pro Blackwell (sm_120); TRITON_ATTN backend works

### ISSUE THREAD

**Issue by SagarSingh131** [vllm-project/vllm#41651]
### Your current environment

<details>
<summary>Environment for the FP8 KV gibberish bug repro</summary>

```text
==============================
        System Info
==============================
OS / Container               : nvcr.io/nvidia/pytorch:26.03-py3 (Ubuntu)
NVIDIA driver                : 580.105.08
CUDA toolkit (host/container): 13.0

==============================
       PyTorch Info
==============================
PyTorch version              : 2.11.0a0+a6c236b9fd.nv26.3.46836102 (NGC build)
Is debug build               : No

==============================
      Python Environment
==============================
Python version               : 3.12.x (container default)

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : Yes
CUDA runtime version         : 13.0
GPU models and configuration : NVIDIA RTX PRO 6000 Blackwell Server Edition
                                arch: GB202 (sm_120), 96 GB VRAM
HIP runtime version          : N/A
MIOpen runtime version       : N/A

==============================
Versions of relevant libraries
==============================
[pip3] torch==2.11.0a0+a6c236b9fd.nv26.3.46836102
[pip3] flashinfer-python==0.6.8.post1
[pip3] flashinfer-cubin==0.6.8.post1
[pip3] transformers==4.57.3
[pip3] compressed-tensors==0.15.0.1
[pip3] triton==3.4.x  (container default)

==============================
         vLLM Info
==============================
vLLM Version                 : 0.19.2rc1 (custom build from source, editable install)
vLLM Build Flags:
  CUDA Archs: sm_120 (Blackwell); ROCm: Disabled

==============================
     Notes
==============================
- H100 and B200 are unaffected with the same model + flags.
```

</details>

## Summary

On **RTX 6000 Pro Blackwell** (sm_120), vLLM with **FlashInfer** attention,
`--kv-cache-dtype fp8`, and the [`poolside/Laguna-XS.2-FP8`](https://huggingface.co/poolside/Laguna-XS.2-FP8) checkpoint (compressed-tensors
`fp8_w8a8` weights with FP8 KV cache via `kv_cache_scheme`) produces
**completely random output** on long prompts when `torch.compile` and CUDA
graphs are enabled.

Switching attention backend to `TRITON_ATTN` while holding everything else
constant produces a **perfect coherent answer** to the same prompt. The bug
is therefore localized to **FlashInfer's interaction with FP8 KV cache + CUDA
graphs on sm_120**.

The bug requires all of:

1. FP8 KV cache enabled (`--kv-cache-dtype fp8`)
2. `torch.compile` and CUDA graphs both enabled (i.e., not `--enforce-eager`)
3. **FlashInfer** attention backend (default on sm_120 in this build)
4. RTX 6000 Pro Blackwell (sm_120). H100 and B200 are clean under the same config.

Short prompts (~30 tokens) are coherent on Blackwell across every config
tested — the bug only manifests at longer context.

## Hardware matrix

| Hardware | Backend | FP8 KV + compile + cudagraphs | Result |
|---|---|---|---|
| RTX 6000 Pro Blackwell (sm_120) | **FlashInfer** | ON | 🔴 Random tokens, no relation to prompt |
| RTX 6000 Pro Blackwell (sm_120) | **TRITON_ATTN** | ON | ✅ Coherent (perfect answer) |
| H100 | FlashInfer | ON | ✅ Coherent |
| B200 | FlashInfer | ON | ✅ Coherent |

## Reproduction

**Setup**

- Hardware: NVIDIA RTX PRO 6000 Blackwell Server Edition (GB202, sm_120, 96 GB).
- Model: [`poolside/Laguna-XS.2-FP8`](https://huggingface.co/poolside/Laguna-XS.2-FP8)
  — a compressed-tensors `fp8_w8a8` weights checkpoint with FP8 KV cache
  configured via the checkpoint's `kv_cache_scheme` (also resolves to `fp8`
  when launched with `--kv-cache-dtype fp8`).
- Prompt: ~2,710-token long-context filler ending in *"Question: Which Apollo
  mission first landed humans on the Moon, and who were its three crew
  members? Answer in one sentence."* Sampling: `temperature=0`.

**Compile / cudagraph configuration that triggers the bug** (from engine init log)

```text
compilation_config.mode = VLLM_COMPILE (3)
cudagraph_mode          = FULL_AND_PIECEWISE
custom_ops              = ['+quant_fp8', 'none', '+quant_fp8']
splitting_ops           include unified_attention_with_output, unified_kv_cache_update
pass_config             fuse_norm_quant=True, fuse_act_quant=True
enforce_eager           = False
kv_cache_dtype          = fp8
```

**A/B server commands** (only `--attention-backend` differs)

```bash
# Reproduces the bug
vllm serve poolside/Laguna-XS.2-FP8 \
    --served-model-name laguna \
    --tensor-parallel-size 1 --trust-remote-code --dtype bfloat16 \
    --kv-cache-dtype fp8 --max-model-len 4096 --gpu-memory-utilization 0.85 \
    --attention-backend FLASHINFER

# Same command, swap backend → bug gone
vllm serve poolside/Laguna-XS.2-FP8 \
    --served-model-name laguna \
    --tensor-parallel-size 1 --trust-remote-code --dtype bfloat16 \
    --kv-cache-dtype fp8 --max-model-len 4096 --gpu-memory-utilization 0.85 \
    --attention-backend TRITON_ATTN
```

Engine init for each leg confirms the named backend was actually loaded:

```text
# FlashInfer leg:
[cuda.py:308] Using AttentionBackendEnum.FLASHINFER backend.

# TRITON_ATTN leg:
[cuda.py:308] Using AttentionBackendEnum.TRITON_ATTN backend.
```

(`TRTLLM`-Gen requires `tcgen05`, which sm_120 lacks, so it is not in the
potential-backends list on this device. The only backends actually available
on sm_120 in this build are FLASHINFER and TRITON_ATTN. Note also that the
`VLLM_ATTENTION_BACKEND` environment variable is no-op'd in vLLM 0.19 — the
`--attention-backend` CLI flag is the working override path.)

**LONG-prompt outputs** (everything else identical)

| `--attention-backend` | Output |
|---|---|
| FLASHINFER  | `" purdude head Magnus cz Pronoyer sizeanu laidazo behind puraldo headko behindMateoki behindmillion Evanzlerpustrap laid teststraposzCertificate laidido protectedaldo lay PIEaldozlerisure hid laidaute laid laid laid laidFld h..."` |
| TRITON_ATTN | `"The Apollo 11 mission first landed humans on the Moon on July 20, 1969, with crew members Neil Armstrong, Buzz Aldrin, and Michael Collins."` |

Short prompts (~30 tokens, e.g. *"What is 17 × 23?"*) return the correct
`391` in both configs.

## Eager-mode behavior (FYI, not part of this bug)

With `--enforce-eager` (which disables `torch.compile` + cudagraphs),
FlashInfer no longer produces random tokens but does enter a partial-recall
repetitive loop on the same long prompt:

```text
"...The Michael Collins Apollo 1969. The program Michael Armstrong Apollo 1968..."
```

This appears to be a separate, milder long-context model issue that is
unmasked once the FlashInfer random-output bug is removed. It is not the
focus of this report.

## Expected behavior

FlashInfer attention + FP8 KV cache + `torch.compile` + CUDA graphs should
produce coherent output on Blackwell-class hardware, matching the H100/B200
behavior under the same model and config — and matching what `TRITON_ATTN`
produces on the same RTX 6000 Pro Blackwell box.

## Workarounds (any one is sufficient)

1. **Switch attention backend with `--attention-backend TRITON_ATTN`** — least
   invasive; FP8 KV + compile + cudagraphs all stay on, and output becomes
   coherent.
2. Disable `torch.compile` + CUDA graphs with `--enforce-eager` (changes the
   symptom from random tokens to a repetitive loop — see "Eager-mode behavior"
   above; doesn't fully restore correctness on long prompts).
3. Disable FP8 KV cache (fall back to default KV dtype).

(1) is the recommended workaround.

## Additional observations

- Bug is gated on long context: short prompts are clean even on affected
  hardware with the broken config.
- Hardware-gated to RTX 6000 Pro Blackwell (sm_120); H100 and B200 are clean
  under the same FlashInfer-default config.
- Container ships `flashinfer-python==0.6.8.post1`. Confirmed reproducing on
  `flashinfer-python==0.6.9` + `flashinfer-cubin==0.6.9` as well (latest as
  of 2026-05-04, force-installed above the `0.6.8.post1` pinned by vLLM 0.19).
  So the bug is not fixed in any released FlashInfer to date.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by msunda17** [comment]
The random output with FlashInfer + FP8 KV cache + CUDA graphs is a classic symptom of silent memory routing issues.

ImpactArbiter is an OSS CLI I built that traps KV-cache/attention-mapping code in a PyTorch autograd graph and detects gradient divergence against a pure Oracle, which is precisely the kind of bug that survives unit tests but breaks under graphs/quantization.

If you can share a minimal repro kernel or mapping function, I’m happy to run it through the tool and post the divergence map here. Repo: https://github.com/msunda17/impactarbiter-cli

**Comment by zhandaz** [comment]
In my reproduction, the TRITON attn backend also didn't work out (on both GB200 and B6000 Pro). I am not sure why TRITON backend and BF16 kv cache would work. I guess the original single prompt test might be a bit random.

Root cause turned out to be two independent bugs that both fire on this model and hardware.

1. **#42580**: `triton_unified_attention` was not applying `k_scale`/`v_scale` on the FP8-Q + FP8 per-tensor KV path, producing wrong output on TRITON_ATTN.
2. **#42650**: both `FlashInferMetadataBuilder` and `TritonAttentionMetadataBuilder` were sourcing `num_qo_heads` / `num_heads_q` from the model-wide `model_config.get_num_attention_heads(...)`. For Laguna XS.2, which sets `num_attention_heads_per_layer = [48, 64, 64, 64, ...]`, the per-layer 64 ends up at the kernel via `q.shape[1]` but the builder allocated for 48. On FlashInfer that under-allocates the plan-time tile budget and the kernel stops writing output at row `total_seq × 48 / 64` (page-aligned), so the tail of the prompt comes back as zeros and the next-token logits are garbage. On Triton the same mismatch makes the kernel write past the typed end of the pre-allocated `softmax_segm_*` scratch and corrupts the adjacent `slot_mapping`.

Why TRTLLM-Gen on sm_100 (GB200) didn't hit it: that path doesn't take `num_qo_heads` at plan time: it reads heads from `query.shape[1]` at runtime. TRTLLM-Gen requires sm_100-family features and isn't dispatched on sm_120, so sm_120 falls back to the native FlashInfer prefill that has the plan/runtime split.

With both PRs applied, verified coherent output on RTX PRO 6000 Blackwell (sm_120) with `poolside/Laguna-XS.2-FP8`, `--kv-cache-dtype fp8`, the same 3597-token Apollo prompt from the bug body, across FlashInfer and TRITON_ATTN, TP=1 and TP=2, and both `--enforce-eager` and full CUDA graphs. gsm8k 100-question accuracy 0.89–0.92 across the matrix, no invalid responses.

Please have a try and let me know if this fixes your issue on your end!

**Comment by DomBrown** [comment]
Seems similar to my fix here #42080 but I also added Q scale != 1 support at the same time. I can confirm this fixes the triton backend in my local tests on RTX Pro 6000

**Comment by DomBrown** [comment]
I appear to have also arrived at the same fix for FlashInfer, but didn't push it yet :)

**Comment by zhandaz** [comment]
> Seems similar to my fix here https://github.com/vllm-project/vllm/pull/42080 but I also added Q scale != 1 support at the same time. I can confirm this fixes the triton backend in my local tests on RTX Pro 6000

Nice! I should have found this earlier, this could save a lot of time for me!
